xarray/tests/test_backends_datatree.py

```diff
@@ -414,112 +446,187 @@ def test_phony_dims_warning(self, tmpdir) -> None:
-        store = ZipStore(filepath)
-        original_dt.to_zarr(store)
+        store = ZipStore(filepath, mode="w")
+        original_dt.to_zarr(store, zarr_format=zarr_format)
 
         with open_datatree(store, engine="zarr") as roundtrip_dt:  # type: ignore[arg-type, unused-ignore]
             assert_equal(original_dt, roundtrip_dt)
 
-    def test_to_zarr_not_consolidated(self, tmpdir, simple_datatree):
+    def test_to_zarr_not_consolidated(self, tmpdir, simple_datatree, zarr_format):
         filepath = tmpdir / "test.zarr"
         zmetadata = filepath / ".zmetadata"
         s1zmetadata = filepath / "set1" / ".zmetadata"
         filepath = str(filepath)  # casting to str avoids a pathlib bug in xarray
         original_dt = simple_datatree
-        original_dt.to_zarr(filepath, consolidated=False)
+        original_dt.to_zarr(filepath, consolidated=False, zarr_format=zarr_format)
         assert not zmetadata.exists()
         assert not s1zmetadata.exists()
 
         with pytest.warns(RuntimeWarning, match="consolidated"):
             with open_datatree(filepath, engine="zarr") as roundtrip_dt:
                 assert_equal(original_dt, roundtrip_dt)
 
-    def test_to_zarr_default_write_mode(self, tmpdir, simple_datatree):
+    def test_to_zarr_default_write_mode(self, tmpdir, simple_datatree, zarr_format):
+        simple_datatree.to_zarr(str(tmpdir), zarr_format=zarr_format)
+
         import zarr
 
-        simple_datatree.to_zarr(tmpdir)
+        # expected exception type changed in zarr-python v2->v3, see https://github.com/zarr-developers/zarr-python/issues/2821
+        expected_exception_type = (
+            FileExistsError if has_zarr_v3 else zarr.errors.ContainsGroupError
+        )
 
         # with default settings, to_zarr should not overwrite an existing dir
-        with pytest.raises(zarr.errors.ContainsGroupError):
-            simple_datatree.to_zarr(tmpdir)
+        with pytest.raises(expected_exception_type):
+            simple_datatree.to_zarr(str(tmpdir))
 
     @requires_dask
-    def test_to_zarr_compute_false(self, tmpdir, simple_datatree):
+    def test_to_zarr_compute_false(
+        self, tmp_path: Path, simple_datatree: DataTree, zarr_format: Literal[2, 3]
+    ):
         import dask.array as da
 
-        filepath = tmpdir / "test.zarr"
+        storepath = tmp_path / "test.zarr"
         original_dt = simple_datatree.chunk()
-        original_dt.to_zarr(filepath, compute=False)
+        original_dt.to_zarr(str(storepath), compute=False, zarr_format=zarr_format)
+
+        def assert_expected_zarr_files_exist(
+            arr_dir: Path,
+            chunks_expected: bool,
+            is_scalar: bool,
+            zarr_format: Literal[2, 3],
+        ) -> None:
+            """For one zarr array, check that all expected metadata and chunk data files exist."""
+            # TODO: This function is now so complicated that it's practically checking compliance with the whole zarr spec...
```

Comment:
I agree. I think the better test here is to write some dask arrays with random data setting `compute=False` and then assert that you only get fill_value on read. That checks that real data is not written with `compute=False`.

Reply:
Can we just punt this to a follow-up PR? All the detailed checks here are so subtle that I'm worried if I try to refactor it again I'm going to either miss things or just delay this PR much further, and ultimately this works fine.